File: epi_judge_python/tree_level_order.py
from test_framework import generic_test
from collections import deque

def binary_tree_depth_order(tree):
    q = deque()
    q.append((1,tree))
    res = []
    while q:
        depth, curr = q.pop()
        if not curr:
            continue
        q.appendleft((depth + 1, curr.left))
        q.appendleft((depth + 1, curr.right))

        if not res or res[-1][-1][0] < depth:
            res.append([(depth, curr.data)])
        else:
            res[-1].append((depth, curr.data))
    return [[x[1] for x in l] for l in res]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("tree_level_order.py",
                                       "tree_level_order.tsv",
                                       binary_tree_depth_order))

File: epi_judge_java/epi/TreeLevelOrder.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.*;

public class TreeLevelOrder {
  private static class NodeInfo<T> {
    BinaryTreeNode<T> node;
    int depth;

    public NodeInfo(BinaryTreeNode<T> node, int depth){
      this.node = node;
      this.depth = depth;
    }
  }
  @EpiTest(testDataFile = "tree_level_order.tsv")

  public static List<List<Integer>>
  binaryTreeDepthOrder(BinaryTreeNode<Integer> tree) {
    if (tree == null){
      return new ArrayList<>();
    }

    Deque<NodeInfo> queue = new ArrayDeque<>();
    queue.add(new NodeInfo<Integer>(tree, 0));
    List<List<Integer>> res = new ArrayList<>();

    while (!queue.isEmpty()){
      NodeInfo currInfo = queue.poll();
      BinaryTreeNode<Integer> u = currInfo.node;
      int depth = currInfo.depth;

      if (u.left != null){
        queue.add(new NodeInfo<Integer>(u.left, depth + 1));
      }
      if (u.right != null){
        queue.add(new NodeInfo<Integer>(u.right, depth + 1));
      }

      if (res.size() > depth){
        List<Integer> listToUpdate = res.get(depth);

        listToUpdate.add(u.data);
      } else {
        ArrayList<Integer> newList = new ArrayList<Integer>();
        newList.add(u.data);
        res.add(newList);
      }
    }
    return res;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "TreeLevelOrder.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
